contrib/coreos/user-data.example

```diff
@@ -190,58 +209,67 @@ write_files:
+          if [ "$HOST" = "$PUBLIC_IP" ] ; then
+            OSD_ID=$(etcdctl get /deis/store/osds/$PUBLIC_IP)
             break
           fi
-          echo "waiting" && sleep 5
-          TIMEWAITED=$((TIMEWAITED+5))
         done
-        docker stop deis-store-daemon
-        docker exec deis-store-admin ceph osd crush remove osd.$OSD_ID
-        docker exec deis-store-admin ceph auth del osd.$OSD_ID
-        docker exec deis-store-admin ceph osd rm $OSD_ID
-        etcdctl rm /deis/store/osds/$PUBLIC_IP
-        etcdctl rm /deis/store/hosts/$PUBLIC_IP && sleep 10
-        # remove ceph mon
-        docker stop deis-store-monitor || true
-        docker exec deis-store-admin ceph mon remove `hostname -f` # fixme
-        docker stop deis-store-metadata || true
-      fi
+        # if we own an osd and its healthy, try to gracefully remove it
+        if [ ! -z "$OSD_ID" ] && [ ${#OSD_HOSTS[@]} -gt "3" ]; then
+          ADMIN_RUNNING=$(docker inspect --format="{{ .State.Running }}" deis-store-admin)
+          if [ $? -eq 1 ] || [ "$ADMIN_RUNNING" == "false" ]; then
+            echo "deis-store-admin container is required for graceful shutdown"
+            exit 2
+          fi
+          set -e -x -o pipefail
+          CURRENT_STATUS=$(docker exec deis-store-admin ceph health | awk '{print $1}')
+          if [[ "$CURRENT_STATUS" != *"HEALTH_OK"* ]]; then
+            echo "Ceph cluster must be healthy to perform graceful removal"
+            exit 3
+          fi
 
-      # removing the node from etcd
-      NODE=$(etcdctl member list | grep `cat /etc/machine-id` | cut -d ':' -f 1)
-      etcdctl member remove $NODE
+          docker exec deis-store-admin ceph osd out $OSD_ID
+          sleep 30
+          TIMEWAITED=0
+          until [[ $(docker exec deis-store-admin ceph health) == *"HEALTH_OK"* ]]
+          do
+            if [ $TIMEWAITED -gt "1200" ]
+            then
+              echo "ceph graceful removal timeout exceeded"
+              break
+            fi
+            echo "waiting" && sleep 5
+            TIMEWAITED=$((TIMEWAITED+5))
+          done
+          docker stop deis-store-daemon
+          docker exec deis-store-admin ceph osd crush remove osd.$OSD_ID
+          docker exec deis-store-admin ceph auth del osd.$OSD_ID
+          docker exec deis-store-admin ceph osd rm $OSD_ID
+          etcdctl rm /deis/store/osds/$PUBLIC_IP
+          etcdctl rm /deis/store/hosts/$PUBLIC_IP && sleep 10
+          # remove ceph mon
+          docker stop deis-store-monitor || true
+          docker exec deis-store-admin ceph mon remove `hostname -f` # fixme
+          docker stop deis-store-metadata || true
+        fi
+      }
+      etcd_shutdown () {
+        set -e -x -o pipefail
+        # removing the node from etcd
+        NODE=$(etcdctl member list | grep `cat /etc/machine-id` | cut -d ':' -f 1)
```

Comment:
On a vanilla contrib/aws setup you get this output

```
etcdctl member list
16a160897c0bc9ab: name=8cc0e1fea18f4995bc1e693f40f09e3f peerURLs=http://10.21.2.151:2380 clientURLs=http://10.21.2.151:2379
876060a75057748c: name=98660f695d88458ab18a1e0c4f0d70e6 peerURLs=http://10.21.1.116:2380 clientURLs=http://10.21.1.116:2379
9adddc1e295abae3: name=e09bea4147a143bcbb850a0b5fb93cbf peerURLs=http://10.21.1.115:2380 clientURLs=http://10.21.1.115:2379
d070be23d802e98b: name=077417806e3f4cdabf8611c6a2e45163 peerURLs=http://10.21.2.150:2380 clientURLs=http://10.21.2.150:2379
ee94bb3ac2fad3e1: name=35f709ccf298449188c6f365176fbaa5 peerURLs=http://10.21.1.114:2380 clientURLs=http://10.21.1.114:2379
```

Looking at `/run/systemd/system/etcd2.service.d/20-cloudinit.conf` I can see `$ETCD_NAME` is not set.

I'd recommend checking for `$ETCD_NAME` and if that's empty fallback on using `/etc/machine-id` for the matching. Including changing the matching to be `name=<id>` to be a bit more robust. This will make custom `name` and `default` work.